File: Leetcode/tree/637_Average of Levels in Binary Tree/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
from collections import deque
class Solution(object):
    def averageOfLevels(self, root):
        """
        :type root: TreeNode
        :rtype: List[float]
        """
        res = []
        if not root: return res
        dq = deque()
        dq.append(root)
        while dq:
            n = len(dq)
            div = 0.0
            for x in range(0, n):
                temp = dq.popleft()
                div += temp.val
                if temp.left: dq.append(temp.left)
                if temp.right: dq.append(temp.right)
            div = div/n
            res.append(div)
        return res

File: Leetcode/tree/637_Average of Levels in Binary Tree/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        vector<double> res;
        if(!root) return res;
        deque<TreeNode*> dq;
        dq.push_back(root);
        while(!dq.empty()){
            int n = dq.size();
            double div = 0.0;
            for(int i = 0; i<n; i++){
                TreeNode* temp = dq.front(); dq.pop_front();
                div += temp->val;
                if(temp->left) dq.push_back(temp->left);
                if(temp->right) dq.push_back(temp->right);
            }
            div /= n;
            res.push_back(div);
        }
        return res;
    }
};
